lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+//===------- ExistentialTransform.cpp - Transform Existential Args -------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Transform existential parameters to generic ones.
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-existential-transform"
+#include "ExistentialTransform.h"
+#include "swift/AST/GenericEnvironment.h"
+#include "swift/AST/GenericSignatureBuilder.h"
+#include "swift/SIL/OptimizationRemark.h"
+#include "swift/SIL/SILFunction.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/TypeSubstCloner.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/CFG.h"
+#include "swift/SILOptimizer/Utils/Existential.h"
+#include "swift/SILOptimizer/Utils/Generics.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+#include "swift/SILOptimizer/Utils/SILOptFunctionBuilder.h"
+#include "swift/SILOptimizer/Utils/SpecializationMangler.h"
+#include "llvm/ADT/SmallVector.h"
+
+using namespace swift;
+
+/// Create a SILCloner for Existential Specilizer.
+namespace {
+class ExistentialSpecializerCloner
+    : public TypeSubstCloner<ExistentialSpecializerCloner,
+                             SILOptFunctionBuilder> {
+  using SuperTy =
+      TypeSubstCloner<ExistentialSpecializerCloner, SILOptFunctionBuilder>;
+  friend class SILInstructionVisitor<ExistentialSpecializerCloner>;
+  friend class SILCloner<ExistentialSpecializerCloner>;
+
+  SILFunction *OrigF;
+  llvm::SmallVector<ArgumentDescriptor, 4> &ArgumentDescList;
+  llvm::SmallDenseMap<int, GenericTypeParamType *> &Arg2GenericTypeMap;
+  llvm::SmallDenseMap<int, ExistentialTransformArgumentDescriptor>
+      &ExistentialArgDescriptor;
+
+protected:
+  void postProcess(SILInstruction *Orig, SILInstruction *Cloned) {
+    SILClonerWithScopes<ExistentialSpecializerCloner>::postProcess(Orig,
+                                                                   Cloned);
+  }
+
+public:
+  ExistentialSpecializerCloner(
+      SILFunction *OrigF, SILFunction *NewF, SubstitutionMap &Subs,
```

Comment:
Unless you're returning the SubstitutionMap by reference, pass it by value -- it's a single pointer so its clean to pass by value